File: CS/AgendaView/Agenda/AgendaViewControl.ascx.cs
using DevExpress.Web;
using DevExpress.Web.ASPxScheduler;
using DevExpress.XtraScheduler;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace AgendaView
{
    public partial class AgendaViewControl : System.Web.UI.UserControl
    {
#region PageEvents
        protected void Page_Load(object sender, EventArgs e)
        {
                InitializeGridControlResources();
                BindAppointmentsGrid();
        }
#endregion
#region Properties
        public ASPxScheduler OwnerScheduler { get; set; }
        public bool ShowResources
        {
            get
            {
                if (Session["ShowResources"] == null)
                    Session["ShowResources"] = false;

                return (bool)Session["ShowResources"];
            }
            set
            {
                Session["ShowResources"] = value;
            }
        }
        public object SelectedResourceId
        {
            get { 
                if (Session["selectedResourceId"] == null)
                    Session["selectedResourceId"] = GridControlResources.GetCardValues(GridControlResources.FocusedCardIndex, "Id");

                return Session["selectedResourceId"];
            }
            set
            {
                Session["selectedResourceId"] = value;
            }
        }
        public TimeInterval SelectedInterval
        {
            get
            {
                if (Session["selectedInterval"] == null)
                {
                    DateTime selectedIntervalStart = OwnerScheduler.ActiveView.GetVisibleIntervals().Start;
                    DateTime intervalStart = new DateTime(selectedIntervalStart.Year, selectedIntervalStart.Month, 1);
                    TimeInterval interval = new TimeInterval(intervalStart, intervalStart.AddMonths(1));
                    Session["selectedInterval"] = interval;
                }

                TimeInterval ti = Session["selectedInterval"] as TimeInterval;
                return ti;
            }
            set
            {
                Session["selectedInterval"] = value;
            }
        }
        public bool Active
        {
            get
            {
                return this.Visible;
            }
        }
#endregion
#region methods
        void InitializeGridControlResources()
        {
            GridControlResources.DataSource = AgendaViewDataGenerator.GenerateResourcesCollection(OwnerScheduler.Storage);
            GridControlResources.DataBind();
        }
         void InitializeGridControlAppointments()
        {
            GridControlAppointments.DataSource = AgendaViewDataGenerator.GenerateAgendaAppointmentCollection(OwnerScheduler.Storage, SelectedResourceId);
            GridControlAppointments.DataBind();
        }
        private void BindAppointmentsGrid()
        {
            AgendaViewDataGenerator.SelectedInterval = SelectedInterval;
            InitializeGridControlAppointments();
            GenerateAgndaViewCaption();
            GridControlAppointments.ExpandAll();
        }
        public void ReloadData()
        {
            BindAppointmentsGrid();
        }
        private void GenerateAgndaViewCaption()
        {
            string currentMonth = SelectedInterval.Start.ToString("MMMM-yyyy");
            int rowCount = GridControlAppointments.VisibleRowCount;
            if (rowCount == 0)
            {
                GridControlAppointments.Caption = currentMonth + " (no data)";
            }
            else if (ShowResources)
            {
                object resourceCaption = GridControlResources.GetCardValuesByKeyValue(SelectedResourceId, "AgendaResourceName");
                if (resourceCaption == null)
                    GridControlAppointments.Caption = currentMonth;
                else
                    GridControlAppointments.Caption = currentMonth + " (" + resourceCaption + ")";

            }
            else
                GridControlAppointments.Caption = currentMonth;
        }
        private void GoToNextMonth()
        {
            SelectedInterval = new TimeInterval(SelectedInterval.End, SelectedInterval.End.AddMonths(1));
            AgendaViewDataGenerator.SelectedInterval = SelectedInterval;
            InitializeGridControlAppointments();
            GenerateAgndaViewCaption();
            GridControlAppointments.ExpandAll();
        }
        private void GoToPreviousMonth()
        {
            SelectedInterval = new TimeInterval(SelectedInterval.Start.AddMonths(-1), SelectedInterval.Start);
            AgendaViewDataGenerator.SelectedInterval = SelectedInterval;
            InitializeGridControlAppointments();
            GenerateAgndaViewCaption();
            GridControlAppointments.ExpandAll();
        }
        private void GoToSpecificDate(DateTime date)
        {
            DateTime intervalStart = new DateTime(date.Year, date.Month, 1);
            TimeInterval interval = new TimeInterval(intervalStart, intervalStart.AddMonths(1));
            SelectedInterval = interval;
            AgendaViewDataGenerator.SelectedInterval = SelectedInterval;
            InitializeGridControlAppointments();
            GenerateAgndaViewCaption();
            GridControlAppointments.ExpandAll();
        }
#endregion
#region events
        protected void GridControlAppointments_CustomUnboundColumnData(object sender, DevExpress.Web.ASPxGridViewColumnDataEventArgs e)
        {
            if (e.Column.Name != "gridColumnStatus") return;

            if (e.IsGetData)
            {
                object status = e.GetListSourceFieldValue("AgendaStatus");
                if (status != null)
                {
                    AppointmentStatus aptStatus = status as AppointmentStatus;
                    e.Value = aptStatus.Color;
                }
            }
        }
        protected void GridControlAppointments_CustomCallback(object sender, DevExpress.Web.ASPxGridViewCustomCallbackEventArgs e)
        {
            if (!GridControlAppointments.IsCallback) return;

            string[] parameters = e.Parameters.Split(';');
            if (parameters.Length != 2) return;

            string commandName = parameters[0];
            string value = parameters[1];
            switch (commandName)
            {
                case "SelectedResourceCommand":
                {
                    SelectedResourceId = Convert.ToInt32(value);
                    if (Convert.ToInt32(SelectedResourceId) == -1)
                        ShowResources = false;
                    else
                        ShowResources = true;
                    BindAppointmentsGrid();
                    break;
                }
                case "SelectedInterval":
                {
                    if (calendar.Value != null)
                    {
                        DateTime dt = Convert.ToDateTime(calendar.Value);
                        GoToSpecificDate(dt);
                    }
                    break;
                }
            }
           
        }
        protected void GridControlAppointments_FillContextMenuItems(object sender, DevExpress.Web.ASPxGridViewContextMenuEventArgs e)
        {
            if (e.MenuType == DevExpress.Web.GridViewContextMenuType.Footer) return;

            e.Items.Clear();
            ContextMenuHelper.FillMenuItems(e.Items, e.MenuType, ShowResources);
        }
        protected void GridControlAppointments_ContextMenuItemClick(object sender, DevExpress.Web.ASPxGridViewContextMenuItemClickEventArgs e)
        {
            switch (e.Item.Name)
            {
                case "GoToNextMonth":
                    GoToNextMonth();
                    break;
                case "GoToPreviousMonth":
                    GoToPreviousMonth();
                    break;
            }
        }
        protected void GridControlAppointments_HtmlRowPrepared(object sender, ASPxGridViewTableRowEventArgs e)
        {
            if (e.RowType != GridViewRowType.Data) return;

            object label = e.GetValue("AgendaLabel");
            if (label != null)
                e.Row.BackColor = (Color)label;
        }
#endregion
    }
}

File: VB/AgendaView/Agenda/AgendaViewControl.ascx.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Web
Imports DevExpress.Web.ASPxScheduler
Imports DevExpress.XtraScheduler
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Drawing
Imports System.Linq
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.WebControls

Namespace AgendaView
	Partial Public Class AgendaViewControl
		Inherits System.Web.UI.UserControl
#Region "PageEvents"
		Protected Sub Page_Load(ByVal sender As Object, ByVal e As EventArgs)
				InitializeGridControlResources()
				BindAppointmentsGrid()
		End Sub
#End Region
#Region "Properties"
		Private privateOwnerScheduler As ASPxScheduler
		Public Property OwnerScheduler() As ASPxScheduler
			Get
				Return privateOwnerScheduler
			End Get
			Set(ByVal value As ASPxScheduler)
				privateOwnerScheduler = value
			End Set
		End Property
		Public Property ShowResources() As Boolean
			Get
				If Session("ShowResources") Is Nothing Then
					Session("ShowResources") = False
				End If

				Return CBool(Session("ShowResources"))
			End Get
			Set(ByVal value As Boolean)
				Session("ShowResources") = value
			End Set
		End Property
		Public Property SelectedResourceId() As Object
			Get
				If Session("selectedResourceId") Is Nothing Then
					Session("selectedResourceId") = GridControlResources.GetCardValues(GridControlResources.FocusedCardIndex, "Id")
				End If

				Return Session("selectedResourceId")
			End Get
			Set(ByVal value As Object)
				Session("selectedResourceId") = value
			End Set
		End Property
		Public Property SelectedInterval() As TimeInterval
			Get
				If Session("selectedInterval") Is Nothing Then
					Dim selectedIntervalStart As DateTime = OwnerScheduler.ActiveView.GetVisibleIntervals().Start
					Dim intervalStart As New DateTime(selectedIntervalStart.Year, selectedIntervalStart.Month, 1)
					Dim interval As New TimeInterval(intervalStart, intervalStart.AddMonths(1))
					Session("selectedInterval") = interval
				End If

				Dim ti As TimeInterval = TryCast(Session("selectedInterval"), TimeInterval)
				Return ti
			End Get
			Set(ByVal value As TimeInterval)
				Session("selectedInterval") = value
			End Set
		End Property
		Public ReadOnly Property Active() As Boolean
			Get
				Return Me.Visible
			End Get
		End Property
#End Region
#Region "methods"
		Private Sub InitializeGridControlResources()
			GridControlResources.DataSource = AgendaViewDataGenerator.GenerateResourcesCollection(OwnerScheduler.Storage)
			GridControlResources.DataBind()
		End Sub
		 Private Sub InitializeGridControlAppointments()
			GridControlAppointments.DataSource = AgendaViewDataGenerator.GenerateAgendaAppointmentCollection(OwnerScheduler.Storage, SelectedResourceId)
			GridControlAppointments.DataBind()
		 End Sub
		Private Sub BindAppointmentsGrid()
			AgendaViewDataGenerator.SelectedInterval = SelectedInterval
			InitializeGridControlAppointments()
			GenerateAgndaViewCaption()
			GridControlAppointments.ExpandAll()
		End Sub
		Public Sub ReloadData()
			BindAppointmentsGrid()
		End Sub
		Private Sub GenerateAgndaViewCaption()
			Dim currentMonth As String = SelectedInterval.Start.ToString("MMMM-yyyy")
			Dim rowCount As Integer = GridControlAppointments.VisibleRowCount
			If rowCount = 0 Then
				GridControlAppointments.Caption = currentMonth & " (no data)"
			ElseIf ShowResources Then
				Dim resourceCaption As Object = GridControlResources.GetCardValuesByKeyValue(SelectedResourceId, "AgendaResourceName")
				If resourceCaption Is Nothing Then
					GridControlAppointments.Caption = currentMonth
				Else
					GridControlAppointments.Caption = currentMonth & " (" & resourceCaption & ")"
				End If

			Else
				GridControlAppointments.Caption = currentMonth
			End If
		End Sub
		Private Sub GoToNextMonth()
			SelectedInterval = New TimeInterval(SelectedInterval.End, SelectedInterval.End.AddMonths(1))
			AgendaViewDataGenerator.SelectedInterval = SelectedInterval
			InitializeGridControlAppointments()
			GenerateAgndaViewCaption()
			GridControlAppointments.ExpandAll()
		End Sub
		Private Sub GoToPreviousMonth()
			SelectedInterval = New TimeInterval(SelectedInterval.Start.AddMonths(-1), SelectedInterval.Start)
			AgendaViewDataGenerator.SelectedInterval = SelectedInterval
			InitializeGridControlAppointments()
			GenerateAgndaViewCaption()
			GridControlAppointments.ExpandAll()
		End Sub
		Private Sub GoToSpecificDate(ByVal [date] As DateTime)
			Dim intervalStart As New DateTime([date].Year, [date].Month, 1)
			Dim interval As New TimeInterval(intervalStart, intervalStart.AddMonths(1))
			SelectedInterval = interval
			AgendaViewDataGenerator.SelectedInterval = SelectedInterval
			InitializeGridControlAppointments()
			GenerateAgndaViewCaption()
			GridControlAppointments.ExpandAll()
		End Sub
#End Region
#Region "events"
		Protected Sub GridControlAppointments_CustomUnboundColumnData(ByVal sender As Object, ByVal e As DevExpress.Web.ASPxGridViewColumnDataEventArgs)
			If e.Column.Name <> "gridColumnStatus" Then
				Return
			End If

			If e.IsGetData Then
				Dim status As Object = e.GetListSourceFieldValue("AgendaStatus")
				If status IsNot Nothing Then
					Dim aptStatus As AppointmentStatus = TryCast(status, AppointmentStatus)
					e.Value = aptStatus.Color
				End If
			End If
		End Sub
		Protected Sub GridControlAppointments_CustomCallback(ByVal sender As Object, ByVal e As DevExpress.Web.ASPxGridViewCustomCallbackEventArgs)
			If (Not GridControlAppointments.IsCallback) Then
				Return
			End If

			Dim parameters() As String = e.Parameters.Split(";"c)
			If parameters.Length <> 2 Then
				Return
			End If

			Dim commandName As String = parameters(0)
			Dim value As String = parameters(1)
			Select Case commandName
				Case "SelectedResourceCommand"
					SelectedResourceId = Convert.ToInt32(value)
					If Convert.ToInt32(SelectedResourceId) = -1 Then
						ShowResources = False
					Else
						ShowResources = True
					End If
					BindAppointmentsGrid()
					Exit Select
				Case "SelectedInterval"
					If calendar.Value IsNot Nothing Then
						Dim dt As DateTime = Convert.ToDateTime(calendar.Value)
						GoToSpecificDate(dt)
					End If
					Exit Select
			End Select

		End Sub
		Protected Sub GridControlAppointments_FillContextMenuItems(ByVal sender As Object, ByVal e As DevExpress.Web.ASPxGridViewContextMenuEventArgs)
			If e.MenuType = DevExpress.Web.GridViewContextMenuType.Footer Then
				Return
			End If

			e.Items.Clear()
			ContextMenuHelper.FillMenuItems(e.Items, e.MenuType, ShowResources)
		End Sub
		Protected Sub GridControlAppointments_ContextMenuItemClick(ByVal sender As Object, ByVal e As DevExpress.Web.ASPxGridViewContextMenuItemClickEventArgs)
			Select Case e.Item.Name
				Case "GoToNextMonth"
					GoToNextMonth()
				Case "GoToPreviousMonth"
					GoToPreviousMonth()
			End Select
		End Sub
		Protected Sub GridControlAppointments_HtmlRowPrepared(ByVal sender As Object, ByVal e As ASPxGridViewTableRowEventArgs)
			If e.RowType <> GridViewRowType.Data Then
				Return
			End If

			Dim label As Object = e.GetValue("AgendaLabel")
			If label IsNot Nothing Then
				e.Row.BackColor = CType(label, Color)
			End If
		End Sub
#End Region
	End Class
End Namespace